network/testing.go

```diff
@@ -20,117 +21,250 @@ import (
-			src, dst := srvs[i], srvs[i+1]
-			doneCh <- src.Join(dst.AddrInfo(), 10*time.Second)
-		}(i)
+	return returnAddrs, nil
+}
+
+func createServers(
+	count int,
+	paramsMap map[int]*CreateServerParams,
+) ([]*Server, error) {
+	servers := make([]*Server, count)
+	if paramsMap == nil {
+		paramsMap = map[int]*CreateServerParams{}
 	}
 
-	for i := 0; i < len(srvs)/2; i++ {
-		err := <-doneCh
-		if err != nil {
-			t.Fatal(err)
+	for i := 0; i < count; i++ {
+		server, createErr := CreateServer(paramsMap[i])
+		if createErr != nil {
+			return nil, createErr
 		}
+
+		servers[i] = server
 	}
+
+	return servers, nil
+}
+
+type CreateServerParams struct {
+	ConfigCallback func(c *Config)      // Additional logic that needs to be executed on the configuration
+	ServerCallback func(server *Server) // Additional logic that needs to be executed on the server before starting
+	Logger         hclog.Logger
 }
 
-func getTestConfig(callback func(c *Config)) *Config {
+var (
+	emptyParams = &CreateServerParams{}
+)
+
+func CreateServer(params *CreateServerParams) (*Server, error) {
 	cfg := DefaultConfig()
-	cfg.Addr.Port = int(atomic.AddUint64(&initialPort, 1))
+	port, portErr := tests.GetFreePort()
+	if portErr != nil {
+		return nil, fmt.Errorf("unable to fetch free port, %v", portErr)
+	}
+	cfg.Addr.Port = port
 	cfg.Chain = &chain.Chain{
 		Params: &chain.Params{
 			ChainID: 1,
 		},
 	}
 
-	logger := hclog.NewNullLogger()
+	if params == nil {
+		params = emptyParams
+	}
 
-	if callback != nil {
-		callback(cfg)
+	if params.ConfigCallback != nil {
+		params.ConfigCallback(cfg)
 	}
 
-	secretsManager, _ := local.SecretsManagerFactory(
+	if params.Logger == nil {
+		params.Logger = hclog.NewNullLogger()
+	}
+
+	secretsManager, factoryErr := local.SecretsManagerFactory(
 		nil,
 		&secrets.SecretsManagerParams{
-			Logger: logger,
+			Logger: params.Logger,
 			Extra: map[string]interface{}{
 				secrets.Path: cfg.DataDir,
 			},
 		},
 	)
+	if factoryErr != nil {
+		return nil, factoryErr
+	}
 
 	cfg.SecretsManager = secretsManager
 
-	return cfg
+	server, err := NewServer(params.Logger, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if params.ServerCallback != nil {
+		params.ServerCallback(server)
+	}
+
+	startErr := server.Start()
+
+	return server, startErr
 }
+
+// MeshJoin is a helper method for joining all the passed in servers into a mesh
+func MeshJoin(servers ...*Server) []error {
+	if len(servers) < 2 {
+		return nil
+	}
+
+	// Join errors are used to gather all errors that happen
+	// inside the go routines, so they can be handled when they finish
+	joinErrors := make([]error, 0)
+	var joinErrorsLock sync.Mutex
+
+	appendJoinError := func(joinErr error) {
+		joinErrorsLock.Lock()
+		joinErrors = append(joinErrors, joinErr)
+		joinErrorsLock.Unlock()
+	}
+
+	numServers := len(servers)
+	var wg sync.WaitGroup
+	for indx := 0; indx < numServers; indx++ {
+		for innerIndx := 0; innerIndx < numServers; innerIndx++ {
+			if innerIndx != indx {
```

Comment:
Does server[1] needs to try to connect to server[0] even though server[0] has dialed to server[1] already? You might want to change the condition like this:

```go
if innerIndx < indx {
    // Skip dialing
    continue
}

```

Reply:
This is a good catch 💯 

When you initially get dialed by a node, you'll also attempt to form an outbound connection.

I've added this optimization:

2ab4d06